Expose waiter configuration

joguSD opened the pull request:
Adding so we can properly address https://github.com/aws/aws-cli/pull/2761

This will allow `WaiterConfig` to be passed to `waiter.wait()` calls allowing customers to adjust waiter's `maxAttempts` and `delay` settings. This is based on how `PaginationConfig` works. The current keys are formatted to match the waiter models.

```
waiter.wait(WaiterConfig={
    'delay': 5,
    'maxAttempts': 5,
})
```

JordonPhillips commented on botocore/waiter.py:
```diff
@@ -287,9 +287,11 @@ def __init__(self, name, config, operation_method):
+        # pop the invocation specific config
+        config = kwargs.pop('WaiterConfig', {})
+        sleep_amount = config.get('delay', self.config.delay)
+        max_attempts = config.get('maxAttempts', self.config.max_attempts)
```
s/maxAttempts/max_attempts/

joguSD replied:
I did this to be consistent with the `PaginateConfig` formatting. The keys are capitalized words there. The waiter config uses mixed case keys. No matter what the key is called it will break consistency in some way.

kyleknap replied:
Little picky, but the casing should be pascal casing where the first letter should be capitalized. That is how pagination does it.

joguSD replied:
I'm okay with `MaxAttempts` and `Delay`.

kyleknap commented:
Approach looks fine. But I agree with Jordon. We need docs for this.

JordonPhillips commented:
It's also worth considering exposing this in the CLI

joguSD commented:
Updated to pascal casing and added the docs in a similar manner to pagination.
<img width="701" alt="screen shot 2017-08-23 at 10 06 37 am" src="https://user-images.githubusercontent.com/1935727/29628843-97c3701e-87eb-11e7-90d6-1c28a0e086e1.png">

kyleknap commented on botocore/docs/waiter.py:
```diff
@@ -78,6 +80,27 @@ def document_wait_method(section, waiter_name, event_emitter,
+    waiter_config_members['Delay'] = DocumentedShape(
+        name='Delay', type_name='integer',
+        documentation=(
+            '<p>The amount of time in seconds to wait between attempts.</p>'))
```
I know this is in the top level description, but it might be good to say what the defaults are for `Delay` and `MaxAttempts` for these operations as well.

kyleknap commented:
Also make sure you add a changelog entity for this.

joguSD commented:
Added defaults inline.
<img width="671" alt="screen shot 2017-08-25 at 1 57 04 pm" src="https://user-images.githubusercontent.com/1935727/29732629-3bfdda30-899e-11e7-8b0b-85e5b11ded79.png">
